// @1inch/limit-order-protocol-utils/connector/private-key-provider.connector.js
import { signTypedData, SignTypedDataVersion } from '@metamask/eth-sig-util';
import { trim0x } from '../utils/limit-order.utils';
export class PrivateKeyProviderConnector {
    constructor(privateKey, web3Provider) {
        this.privateKey = privateKey;
        this.web3Provider = web3Provider;
    }
    contractEncodeABI(abi, address, methodName, methodParams) {
        const contract = new this.web3Provider.eth.Contract(abi, address === null ? undefined : address);
        return contract.methods[methodName](...methodParams).encodeABI();
    }
    signTypedData(_walletAddress, typedData, 
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    _typedDataHash = '') {
        const result = signTypedData({
            privateKey: Buffer.from(this.privateKey, 'hex'),
            data: typedData,
            version: SignTypedDataVersion.V4,
        });
        return Promise.resolve(result);
    }
    ethCall(contractAddress, callData) {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }
    decodeABIParameter(type, hex) {
        return this.web3Provider.eth.abi.decodeParameter(type, hex);
    }
    decodeABICallParameters(types, callData) {
        const parameters = trim0x(callData).substring(8);
        return this.web3Provider.eth.abi.decodeParameters([...types], parameters);
    }
}
