// @1inch/limit-order-protocol-utils/limit-order-protocol-addresses.const.js
export var ChainId;
(function (ChainId) {
    ChainId[ChainId["ethereumMainnet"] = 1] = "ethereumMainnet";
    ChainId[ChainId["binanceMainnet"] = 56] = "binanceMainnet";
    ChainId[ChainId["polygonMainnet"] = 137] = "polygonMainnet";
    ChainId[ChainId["optimismMainnet"] = 10] = "optimismMainnet";
    ChainId[ChainId["arbitrumMainnet"] = 42161] = "arbitrumMainnet";
    ChainId[ChainId["gnosisMainnet"] = 100] = "gnosisMainnet";
    ChainId[ChainId["avalancheMainnet"] = 43114] = "avalancheMainnet";
    ChainId[ChainId["fantomMainnet"] = 250] = "fantomMainnet";
    ChainId[ChainId["auroraMainnet"] = 1313161554] = "auroraMainnet";
    ChainId[ChainId["klaytnMainnet"] = 8217] = "klaytnMainnet";
    ChainId[ChainId["zkSyncEraMainnet"] = 324] = "zkSyncEraMainnet";
    ChainId[ChainId["baseMainnet"] = 8453] = "baseMainnet";
})(ChainId || (ChainId = {}));
export const limitOrderProtocolAddresses = {
    [ChainId.ethereumMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.binanceMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.polygonMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.optimismMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.arbitrumMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.auroraMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.gnosisMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.avalancheMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.fantomMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.klaytnMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
    [ChainId.zkSyncEraMainnet]: '0x6e2b76966cbd9cf4cc2fa0d76d24d5241e0abc2f',
    [ChainId.baseMainnet]: '0x1111111254eeb25477b68fb85ed929f73a960582',
};
