// @1inch/limit-order-protocol-utils/model/limit-order-protocol.model.js
export const InteractionsFieldsV3 = {
    makerAssetData: 0,
    takerAssetData: 1,
    getMakingAmount: 2,
    getTakingAmount: 3,
    predicate: 4,
    permit: 5,
    preInteraction: 6,
    postInteraction: 7,
    // cuz enum has numeric keys also
};
export const InteractionsFields = {
    makerAssetSuffix: 0,
    takerAssetSuffix: 1,
    makingAmountGetter: 2,
    takingAmountGetter: 3,
    predicate: 4,
    permit: 5,
    preInteraction: 6,
    postInteraction: 7
};
export var LimitOrderProtocolMethodsV3;
(function (LimitOrderProtocolMethodsV3) {
    LimitOrderProtocolMethodsV3["cancelOrder"] = "cancelOrder";
    LimitOrderProtocolMethodsV3["timestampBelow"] = "timestampBelow";
    LimitOrderProtocolMethodsV3["timestampBelowAndNonceEquals"] = "timestampBelowAndNonceEquals";
    LimitOrderProtocolMethodsV3["checkPredicate"] = "checkPredicate";
    LimitOrderProtocolMethodsV3["increaseNonce"] = "increaseNonce";
    LimitOrderProtocolMethodsV3["nonce"] = "nonce";
    LimitOrderProtocolMethodsV3["advanceNonce"] = "advanceNonce";
    LimitOrderProtocolMethodsV3["and"] = "and";
    LimitOrderProtocolMethodsV3["or"] = "or";
    LimitOrderProtocolMethodsV3["eq"] = "eq";
    LimitOrderProtocolMethodsV3["lt"] = "lt";
    LimitOrderProtocolMethodsV3["gt"] = "gt";
    LimitOrderProtocolMethodsV3["nonceEquals"] = "nonceEquals";
    LimitOrderProtocolMethodsV3["arbitraryStaticCall"] = "arbitraryStaticCall";
    LimitOrderProtocolMethodsV3["remaining"] = "remaining";
    LimitOrderProtocolMethodsV3["hashOrder"] = "hashOrder";
    LimitOrderProtocolMethodsV3["isValidSignature"] = "isValidSignature";
})(LimitOrderProtocolMethodsV3 || (LimitOrderProtocolMethodsV3 = {}));
export var LimitOrderProtocolMethods;
(function (LimitOrderProtocolMethods) {
    LimitOrderProtocolMethods["getMakingAmount"] = "getMakingAmount";
    LimitOrderProtocolMethods["getTakingAmount"] = "getTakingAmount";
    LimitOrderProtocolMethods["arbitraryStaticCall"] = "arbitraryStaticCall";
    LimitOrderProtocolMethods["fillOrder"] = "fillOrder";
    LimitOrderProtocolMethods["fillOrderArgs"] = "fillOrderArgs";
    LimitOrderProtocolMethods["cancelOrder"] = "cancelOrder";
    LimitOrderProtocolMethods["permitAndCall"] = "permitAndCall";
    LimitOrderProtocolMethods["increaseEpoch"] = "increaseEpoch";
    LimitOrderProtocolMethods["remainingInvalidatorForOrder"] = "remainingInvalidatorForOrder";
    LimitOrderProtocolMethods["rawRemainingInvalidatorForOrder"] = "rawRemainingInvalidatorForOrder";
    LimitOrderProtocolMethods["epoch"] = "epoch";
    LimitOrderProtocolMethods["checkPredicate"] = "checkPredicate";
    LimitOrderProtocolMethods["advanceNonce"] = "advanceNonce";
    LimitOrderProtocolMethods["increaseNonce"] = "increaseNonce";
    LimitOrderProtocolMethods["hashOrder"] = "hashOrder";
    LimitOrderProtocolMethods["and"] = "and";
    LimitOrderProtocolMethods["or"] = "or";
    LimitOrderProtocolMethods["eq"] = "eq";
    LimitOrderProtocolMethods["lt"] = "lt";
    LimitOrderProtocolMethods["gt"] = "gt";
    LimitOrderProtocolMethods["nonceEquals"] = "nonceEquals";
    LimitOrderProtocolMethods["transferFrom"] = "transferFrom";
})(LimitOrderProtocolMethods || (LimitOrderProtocolMethods = {}));
