// @1inch/solidity-utils/dist/typechain-types/factories/@openzeppelin/contracts/utils/math/Math__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Math__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "MathOverflowedMulDiv",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea2646970667358221220151e2d57d759c5821c8cf48eff0bdfa33c018744a1c44772d758e071865fa8f064736f6c63430008170033";
const isSuperArgs = (xs) => xs.length > 1;
class Math__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.Math__factory = Math__factory;
Math__factory.bytecode = _bytecode;
Math__factory.abi = _abi;
