// @1inch/solidity-utils/dist/typechain-types/factories/@openzeppelin/contracts/utils/Strings__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strings__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "value",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "length",
                type: "uint256",
            },
        ],
        name: "StringsInsufficientHexLength",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea264697066735822122022ac4017e0e6acab4084fc8500e4e9832331ac0ca8659956e3dda366ca5f9cf364736f6c63430008170033";
const isSuperArgs = (xs) => xs.length > 1;
class Strings__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.Strings__factory = Strings__factory;
Strings__factory.bytecode = _bytecode;
Strings__factory.abi = _abi;
