// @1inch/solidity-utils/dist/typechain-types/factories/contracts/tests/mocks/EthReceiverMock__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthReceiverMock__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [],
        name: "EthDepositRejected",
        type: "error",
    },
    {
        stateMutability: "payable",
        type: "receive",
    },
];
const _bytecode = "0x608080604052346013576074908160188239f35b5f80fdfe6080806040523615600e575f80fd5b323314601657005b807f1b10b0f90000000000000000000000000000000000000000000000000000000060049252fdfea2646970667358221220f8f335828d24baadfe1029f8d6af3c2aebfc102d5be7875858c5cc4f074f716864736f6c63430008170033";
const isSuperArgs = (xs) => xs.length > 1;
class EthReceiverMock__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.EthReceiverMock__factory = EthReceiverMock__factory;
EthReceiverMock__factory.bytecode = _bytecode;
EthReceiverMock__factory.abi = _abi;
