// @1inch/limit-order-sdk/dist/cjs/limit-order-contract/native-order-impl.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "NativeOrdersImpl", {
    enumerable: true,
    get: function() {
        return NativeOrdersImpl;
    }
});
var _ethers = require("ethers");
var _NativeOrderImplabijson = /*#__PURE__*/ _interop_require_default(require("../abi/NativeOrderImpl.abi.json"));
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
var NativeOrdersImpl = /*#__PURE__*/ function() {
    "use strict";
    function NativeOrdersImpl(address) {
        _class_call_check(this, NativeOrdersImpl);
        _define_property(this, "address", void 0);
        _define_property(this, "iface", void 0);
        this.address = address;
        this.iface = new _ethers.Interface(_NativeOrderImplabijson.default);
    }
    _create_class(NativeOrdersImpl, [
        {
            key: "cancel",
            value: function cancel(maker, order) {
                return {
                    to: this.address,
                    value: 0n,
                    data: this.iface.encodeFunctionData('cancelOrder', [
                        _object_spread_props(_object_spread({}, order), {
                            maker: maker.toString()
                        })
                    ])
                };
            }
        },
        {
            key: "cancelExpiredOrderByResolver",
            value: function cancelExpiredOrderByResolver(maker, order, rewardLimit) {
                return {
                    to: this.address,
                    value: 0n,
                    data: this.iface.encodeFunctionData('cancelExpiredOrderByResolver', [
                        _object_spread_props(_object_spread({}, order), {
                            maker: maker.toString()
                        }),
                        rewardLimit
                    ])
                };
            }
        }
    ]);
    return NativeOrdersImpl;
}();
