// @1inch/solidity-utils/dist/typechain-types/factories/contracts/libraries/SafeERC20__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SafeERC20__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "ForceApproveFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "Permit2TransferAmountTooHigh",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeDecreaseAllowanceFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeIncreaseAllowanceFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "SafePermitBadLength",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeTransferFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "SafeTransferFromFailed",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea26469706673582212209b2795b4376eab49c6fa5803300dc9a46bca9e1540e0d592ee7a702bd3b6d7f764736f6c63430008170033";
const isSuperArgs = (xs) => xs.length > 1;
class SafeERC20__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.SafeERC20__factory = SafeERC20__factory;
SafeERC20__factory.bytecode = _bytecode;
SafeERC20__factory.abi = _abi;
